core/deployment/src/main/java/io/quarkus/deployment/steps/ConfigBuildSteps.java

```diff
@@ -97,4 +102,20 @@ void nativeServiceProviders(
     RuntimeInitializedClassBuildItem runtimeInitializedClass() {
         return new RuntimeInitializedClassBuildItem(InetRunTime.class.getName());
     }
+
+    @BuildStep(onlyIf = SystemOnlySources.class)
+    void systemOnlySources(BuildProducer<StaticInitConfigBuilderBuildItem> staticInitConfigBuilder,
+            BuildProducer<RunTimeConfigBuilderBuildItem> runTimeConfigBuilder) {
+        staticInitConfigBuilder.produce(new StaticInitConfigBuilderBuildItem(SystemOnlySourcesConfigBuilder.class.getName()));
+        runTimeConfigBuilder.produce(new RunTimeConfigBuilderBuildItem(SystemOnlySourcesConfigBuilder.class.getName()));
+    }
+
+    private static class SystemOnlySources implements BooleanSupplier {
+        ConfigBuildTimeConfig configBuildTimeConfig;
+
+        @Override
+        public boolean getAsBoolean() {
+            return configBuildTimeConfig.systemOnly;
+        }
+    }
```

Comment:
I am a little confused now. This was at least partly driven by what the Quarkus CLI needs. It absolutely needs to read application.properties _at build time_. It just can't/shouldn't look for anything else at runtime. 

But I think I'm probably reading too much into this... This would constrain what is available to static init (which is recorded at build time but run at run time?)

Reply:
Yes, because we do generate the static init / runtime configuration during the build. So, if the configuration is active, we just produce config builders for static and runtime that remove the unwanted config sources.